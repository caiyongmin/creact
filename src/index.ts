import React, {
  useState,
  useEffect,
  useReducer,
  useCallback,
  useMemo,
  useRef,
  useContext,
} from './react/index';
import ReactDOM from './react-dom/index';

export {
  React,
  ReactDOM,
  useState,
  useEffect,
  useReducer,
  useCallback,
  useMemo,
  useRef,
  useContext,
}
