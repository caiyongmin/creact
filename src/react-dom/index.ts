import { render } from './render';
import { diff } from './diff';

export default {
  render,
  diff
};
