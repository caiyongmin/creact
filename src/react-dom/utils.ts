import { VNode } from "../../typings/index";

export function isStrOrNum(val: any): val is string | number {
  return typeof val === 'string' || typeof val === 'number';
}

export function isUnRenderVDom(val: any): val is boolean | undefined | null {
  return typeof val === 'boolean' || val === undefined || val === null;
}

export function isObject(val: any): boolean {
  return typeof val === 'object' && val !== null;
}

export function isFunction(val: any): val is Function {
  return typeof val === 'function';
}

export function getFunctionBody (fn: Function): string {
  const toString = Function.prototype.toString;
  return toString.call(fn).replace(/^[^{]*{\s*/, '').replace(/\s*}[^}]*$/, '')
}

export function isClass(val: any): boolean {
  if (typeof val !== 'function') {
    return false;
  }

  if (/^class[\s{]/.test(val.toString())) {
    return true;
  }

  const fnBody = getFunctionBody(val);
  const result = (/classCallCheck/.test(fnBody)
    || /TypeError\("Cannot call a class as a function"\)/.test(fnBody));

  return result;
}

export function isString(val: any): val is string {
  return typeof val === 'string';
}

export function isSameNodeType(dom: any, vdom: VNode): boolean {
  // 文本节点
  if (typeof vdom === 'string' || typeof vdom === 'number') {
    return dom.nodeType === 3;
  }

  // DOM 节点
  if (typeof vdom === 'object' && vdom !== null && typeof vdom.type === 'string') {
    return dom.nodeName.toLowerCase() === vdom.type.toLowerCase();
  }

  // 组件
  if (dom && dom.__componentInstance && typeof vdom === 'object' && vdom !== null && vdom.type) {
    return dom.__componentInstance.renderVDOM === vdom.type;
  }

  return false;
}

export function shallowCompareObject(
  obj: { [ key: string ]: any },
  compareObj: { [ key: string ]: any },
) {
  if (!isObject(obj) || !isObject(compareObj)) {
    return false;
  }

  const objKeys = Object.keys(obj);
  const compareObjKeys = Object.keys(compareObj);

  if (objKeys.length !== compareObjKeys.length) {
    return false;
  }

  let index = objKeys.length;
  while (index--) {
    const key = objKeys[index];

    if (
      !Object.hasOwnProperty.call(compareObj, key)
      || compareObj[key] !== obj[key]
    ) {
      return false;
    }
  }

  return true;
}
