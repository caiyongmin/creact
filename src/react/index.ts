import Component from './component';
import createElement from './createElement';

export { setDispatcher, useState, useEffect } from './hooks';

export default {
  Component,
  createElement
}
