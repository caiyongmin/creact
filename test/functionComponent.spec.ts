import { createRoot, deleteRoot } from './utils';
import { ReactDOM } from '../src';
import { useStateVNode } from './fixtures/useState';

describe('function component', () => {
  let root: HTMLDivElement;

  beforeEach(() => {
    root = createRoot();
  });

  afterEach(() => {
    deleteRoot(root);
  });

  it('should render correct DOM', () => {
    ReactDOM.render(useStateVNode, root);
    expect(root).toMatchSnapshot();
  });

  it('should render correct DOM after virtual dom diff', () => {
    ReactDOM.render(useStateVNode, root);
    const addButton = root.querySelector('#add') as HTMLButtonElement;
    const result = root.querySelector('#result') as HTMLSpanElement;
    expect(result.innerHTML).toBe('0');
    addButton.click();
    expect(result.innerHTML).toBe("1");
  });
});
