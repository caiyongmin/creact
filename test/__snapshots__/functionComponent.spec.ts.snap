// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function component should render correct DOM 1`] = `
<div
  id="root"
>
  <div>
    <button
      id="add"
    >
      addCount
    </button>
    <button
      id="reset"
    >
      reset
    </button>
    <button
      id="same"
    >
      same
    </button>
    <span
      id="result"
    >
      0
    </span>
  </div>
</div>
`;
